src/rate_limiter/src/lib.rs

```diff
@@ -161,14 +161,21 @@ impl TokenBucket {
     fn auto_replenish(&mut self) {
         // Compute time passed since last refill/update.
         let time_delta = self.last_update.elapsed().as_nanos() as u64;
-        self.last_update = Instant::now();
 
         // At each 'time_delta' nanoseconds the bucket should refill with:
         // refill_amount = (time_delta * size) / (complete_refill_time_ms * 1_000_000)
         // `processed_capacity` and `processed_refill_time` are the result of simplifying above
         // fraction formula with their greatest-common-factor.
         let tokens = (time_delta * self.processed_capacity) / self.processed_refill_time;
-        self.budget = std::cmp::min(self.budget + tokens, self.size);
+
+        // Frequently calling `auto_replenish` will reset `self.last_update` each
+        // time and `tokens` may be a fractional value (0 since it is a `u64`), in
+        // this case no tokens will be replenished. To avoid this we only update
+        // `self.last_update` when `tokens > 0`.
+        if tokens > 0 {
+            self.last_update = Instant::now();
```

Comment:
```suggestion
            self.last_update += Duration::from_nanos((tokens * self.processed_refill_time) / self.processed_capacity);
```

Reply:
can't this division fall into the exact same problem as before?

Reply:
oh, I see now. it doesn't matter. if this division yields zero it means we have replenished zero tokens, right? I think it makes sense.

Reply:
This does fix it yes.

Reply:
which is equivalent to 
```suggestion
            self.last_update += time_delta;
```
right?

Reply:
I think it doesn't really matter. Both cases will appear if we access the rate limiter with extremely high resolution, i.e. ~nanoseconds. This is not normally the case.

Reply:
> This does however mean we are incurring a time debt on every call. self.last_update will gradually fall further behind Instant::now().

should not fall behind.. every call new `time_delta` is `last_update.elapsed()` where `elapsed()` will "cover" what you "lost" last call (I use "" because nothing is actually lost, and making use of `time_delta` is actually more accurate than calling `Instant::now()` again).

Reply:
Yes, I think Adrian is right. If every time we increase by `time_delta` we actually have the resolution of `.elapsed()` (we only miss whatever `.elapsed()` misses), which should be tiny and it should solve the problem.

Reply:
As I noted in https://github.com/firecracker-microvm/firecracker/pull/3370#discussion_r1072359788

> I think something like https://crates.io/crates/governor (more specifically https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm) would help with this but this would be a more significant switch.

Reply:
hmmm, but then, `time_delta` might also include fractions of a token. We would loose that as well. Patrick's calculation discounts the fraction, it takes only the integer part, but it looses some time. In any case:

> I think it doesn't really matter. Both cases will appear if we access the rate limiter with extremely high resolution, > i.e. ~nanoseconds. This is not normally the case.

Reply:
> As I noted in [#3370 (comment)](https://github.com/firecracker-microvm/firecracker/pull/3370#discussion_r1072359788)
> 
> > I think something like https://crates.io/crates/governor (more specifically https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm) would help with this but this would be a more significant switch.

I don't know whether this solves our issue or not, we would need to see their implementation. In general, I would expect _any_ rate limiting implementation to not have infinite resolution.

Reply:
> > This does however mean we are incurring a time debt on every call. self.last_update will gradually fall further behind Instant::now().
> 
> should not fall behind.. every call new `time_delta` is `last_update.elapsed()` where `elapsed()` will "cover" what you "lost" last call

It won't cover all the time loss, since time is lost between taking the time measurement and calculating the number of tokens. This should only be nanoseconds,

Reply:
This "rate-limiter" is very dumb anyway, it works in spikes, trying to hit an **average limit over time**. All the available bucket tokens can be instantly expended (spike well over limit), but then it's forcefully cooldowned for 100ms, then another spike, & so on..

And while simple and stupid, it does the job with good normalized results at scale with very little complexity as opposed to other "smarter" mechanisms :laughing: